### PR TITLE
[core][experimental] Avoid false positives in deadlock detection

### PR DESCRIPTION
**Pull request by kevin85421**
## Why are these changes needed?

Initially, the deadlock detection mechanism was designed based on a DAGNode-based execution schedule, meaning the read, compute, and write operations belonging to the same DAGNode will be executed consecutively.

Then, #46911 changes the execution schedule from DAGNode-based to operation-based to avoid deadlocks caused by NCCL operations. Therefore, there are some false positives for the old deadlock detection mechanism.

With this PR, the deadlock detection focuses only on NCCL operations where both the reader and writer belong to the same actor.

### Future works: Avoid false negatives

```
actor1.f1 ---> actor2.f1
           |
actor1.f2 --
```

In this case, `actor1.f1` and `actor1.f2` have control dependencies between them. If `actor2.f1` reads `actor1.f2` first and then `actor1.f1`, there will be a deadlock. However, this deadlock is not detectable until we have a more granular execution schedule.

## Checks

- [ ] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
    - [ ] I've added any new APIs to the API Reference. For example, if I added a 
           method in Tune, I've added it in `doc/source/tune/api/` under the 
           corresponding `.rst` file.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: python/ray/dag/tests/experimental/test_torch_tensor_dag.py [hunk 1]
```diff
@@ -746,28 +745,6 @@ def test_torch_tensor_nccl_nested_dynamic(ray_start_regular):
     compiled_dag.teardown()
 
 
-@pytest.mark.parametrize("ray_start_regular", [{"num_cpus": 4}], indirect=True)
```
Comment (kevin85421) [inline python/ray/dag/tests/experimental/test_torch_tensor_dag.py:749]: This has already been covered by tests in `test_detect_deadlock_dag.py`.

File: python/ray/dag/compiled_dag_node.py [hunk 2]
```diff
@@ -1620,15 +1512,9 @@ def _is_same_actor(idx1: int, idx2: int) -> bool:
...
             return actor_id_1 == actor_id_2
 
-        graph = defaultdict(GraphNode)
         for idx, task in self.idx_to_task.items():
-            # Add an edge from task_{bind_index} to task_{bind_index+1}
-            # on the same actor.
-            next_task_idx = _get_next_task_idx(task)
-            _add_edge(graph, idx, next_task_idx)
             for downstream_idx in task.downstream_task_idxs:
                 # Add an edge from the writer to the reader.
```
Comment (ruisearch42) [inline python/ray/dag/compiled_dag_node.py:1517]: remove comment

**Review by ruisearch42 — APPROVED** [review]

File: python/ray/dag/compiled_dag_node.py (hunk 2 above)
Comment (kevin85421) [reply python/ray/dag/compiled_dag_node.py:1517]: Oops, good catch

Comment (kevin85421) [reply python/ray/dag/compiled_dag_node.py]: updated https://github.com/ray-project/ray/pull/47912/commits/4941d2cbf52ffb9d55156659c337788c676c6176

**Pull request merged**